presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -327,8 +365,28 @@ void collectDynamicFilters()
 
     private static Set<DynamicFilterId> getLazyDynamicFilters(PlanFragment plan)
     {
-        // lazy dynamic filters cannot be consumed by the same stage where they are produced as it would result in query deadlock
-        return difference(getProducedDynamicFilters(plan.getRoot()), getConsumedDynamicFilters(plan.getRoot()));
+        // To prevent deadlock dynamic filter can be lazy only when:
+        // 1. it's consumed by different stage from where it's produced
+        // 2. or it's produced by replicated join in source stage. In such case an extra
+        //    task is created that will collect dynamic filter and prevent deadlock.
+        Set<DynamicFilterId> interStageDynamicFilters = difference(getProducedDynamicFilters(plan.getRoot()), getConsumedDynamicFilters(plan.getRoot()));
+        return union(interStageDynamicFilters, getSourceStageInnerLazyDynamicFilters(plan));
```

Comment:
`Sets.difference` and `Sets.union` return a "view" which is computed for each iteration and will change when the underlying sets change.  I would make a defensive copy here using `ImmutableSet`.